### PR TITLE
Improved LwjglGraphics.checkOpenGL() output

### PR DESCRIPTION
**Pull request by emanuele3d**
Improved LwjglGraphics.checkOpenGL() to be more informative and specific in case of GPU compatibility problems.

### CODE REVIEW

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java [hunk 1]
```diff
@@ -254,22 +255,87 @@ public Subtexture buildAsset(AssetUri uri, SubtextureData data) {
...
+        boolean[] requiredCapabilities = {
+                GLContext.getCapabilities().OpenGL11,
+                GLContext.getCapabilities().OpenGL12,
+                GLContext.getCapabilities().OpenGL14,
+                GLContext.getCapabilities().OpenGL15,
+                GLContext.getCapabilities().GL_ARB_framebuffer_object,
+                GLContext.getCapabilities().GL_ARB_texture_float,
+                GLContext.getCapabilities().GL_ARB_half_float_pixel,
+                GLContext.getCapabilities().GL_ARB_shader_objects};
+
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: For testing purposes, you might want to add here something like:

    requiredCapabilities[3] = false;
    requiredCapabilities[6] = false;

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java [hunk 2]
```diff
@@ -254,22 +255,87 @@ public Subtexture buildAsset(AssetUri uri, SubtextureData data) {
...
+        String completeErrorMessage = "\n" +
+                "\nThe following OpenGL versions/extensions are required but are not supported by your GPU driver:\n" +
+                "\n" +
+                errorMessage +
+                "\n" +
+                "GPU Information:\n" +
+                "\n";
+
+        if (GLContext.getCapabilities().OpenGL11) {
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: I haven't managed to figure this out, but I've been wondering if having a working GLContext implies the availability of at least OpenGL 1.1.

**Comment by Cervator** [comment]
Nice, this is for #1548 I figure? :-)

Although, how do we even truly test this? I guess we can just fake it with the booleans?

I'll be no help with the GLContext fun for sure though

**Comment by emanuele3d** [comment]
Yes Cervator, it is for #1548. I forgot to mention it, sorry.

I faked it with booleans temporarily added to the code, as suggested above. There are companies that specialize in starting up software on a variety of platforms and configuration, sending back to the developer any generated log and the actual hardware/software descriptions. But I suspect it's a bit on the expensive side...

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java [hunk 3]
```diff
@@ -254,22 +255,85 @@ public Subtexture buildAsset(AssetUri uri, SubtextureData data) {
...
+                GLContext.getCapabilities().OpenGL11,
+                GLContext.getCapabilities().OpenGL12,
+                GLContext.getCapabilities().OpenGL14,
+                GLContext.getCapabilities().OpenGL15,
+                GLContext.getCapabilities().GL_ARB_framebuffer_object,
+                GLContext.getCapabilities().GL_ARB_texture_float,
+                GLContext.getCapabilities().GL_ARB_half_float_pixel,
+                GLContext.getCapabilities().GL_ARB_shader_objects};
+
+        requiredCapabilities[3] = false;
```
Comment (msteiger) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: I guess this line should not be in there.

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java [hunk 4]
```diff
@@ -254,22 +255,85 @@ public Subtexture buildAsset(AssetUri uri, SubtextureData data) {
...
+            if (!requiredCapabilities[index]) {
+                missingCapabilitiesMessage += "    - " + capabilityNames[index] + "\n";
+                canRunTheGame = false;
+            }
         }
 
+        if (!canRunTheGame) {
+            String completeErrorMessage = completeErrorMessage(missingCapabilitiesMessage);
+            logger.error(completeErrorMessage);
+            JOptionPane.showMessageDialog(null, completeErrorMessage, "Required OpenGL version(s) or extension(s) not supported", JOptionPane.ERROR_MESSAGE);
```
Comment (msteiger) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: You need to call AWT methods from the EventDispatchThread using SwingUtilities.invoke.. Otherwise the splashscreen which might run in the background crashes with an IllegalStateEx.

I'd remove this line entirely as it is not the job of `LwjglGraphics` to show nice GUI dialogs. If you

     throw new IllegalStateException(completeErrorMessage);

instead, it will bubble up to `main()` and will be reported by CrashManager (if not headless, CR enabled, etc).

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java [hunk 5]
```diff
@@ -254,22 +255,85 @@ public Subtexture buildAsset(AssetUri uri, SubtextureData data) {
...
+            System.exit(1);
+        }
+    }
+
+    private String completeErrorMessage(String errorMessage) {
+
+        String completeErrorMessage = "\n" +
```
Comment (msteiger) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: I would remove the two newline chars. before the text, as it looks quite odd the in GUI dialog.

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java (hunk 2 above)
Comment (msteiger) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: According to Wikipedia, OpenGL 1.1 was released in 1997. I would assume that every gfx card on the market supports it. If it is not available, the OpenGL drivers (usually shipped with OS and gfx card drivers) are probably not installed.

Then again, I would assume that the `GLContext` cannot be created if the drivers are missing entirely.

Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: I agree.

I just spent some time digging in LWJGL documentation and forum. I'm still not sure 100%, but I think if the line Display.create() in LwjglGraphics.initDisplay() succeeds, *at least* an opengl 1.1 context is available through the GLContext object.

So, in the next commit I've eliminated the checks regarding opengl 1.1 and simplified the code accordingly.

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java (hunk 3 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: D'oh. Sorry about that.

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java (hunk 4 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: I just did as you suggest, in the next commit. I find the stacktrace -after- the formatted part of the error somewhat distracting for the user and not particularly helpful for us either, which is why I called System.exit(1) instead. Any way to suppress it?

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java (hunk 5 above)
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: I've removed the GUI dialog lines as per previous suggestion. So, I'll keep the two newlines chars unless you tell me that you do not like them in the log either... =)

File: engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java [hunk 6]
```diff
@@ -16,11 +16,7 @@
 package org.terasology.engine.subsystem.lwjgl;
 
 import org.lwjgl.LWJGLException;
-import org.lwjgl.opengl.ContextAttribs;
-import org.lwjgl.opengl.Display;
-import org.lwjgl.opengl.GL43;
-import org.lwjgl.opengl.GLContext;
-import org.lwjgl.opengl.KHRDebugCallback;
+import org.lwjgl.opengl.*;
```
Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: IntelliJ keeps transforming this back to an opengl.* statement while checkstyle flags it as a problem. Anything I can do about it?

Comment (Cervator) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: File - Settings - Editor - Code Style - Java - Imports - Class count to use import with '*' :-)

Just increase it to a large number

Probably a thing we can bake into the IntelliJ config via Gradle

Comment (emanuele3d) [inline engine/src/main/java/org/terasology/engine/subsystem/lwjgl/LwjglGraphics.java]: Thank you! Done. I won't submit another commit to fix this unless some other changes are needed. But now the number is at 1000 so, won't happen again.

**Comment by msteiger** [comment]
I think we can merge this.

**Comment by Cervator** [comment]
Tested with a hacked boolean - very nice and clear! Good stuff :-)

```
13:27:52.158 [main] ERROR o.terasology.engine.TerasologyEngine - Failed to initialise Terasology
java.lang.IllegalStateException: 

The following OpenGL versions/extensions are required but are not supported by your GPU driver:

    - OpenGL15

GPU Information:

    Vendor:  ATI Technologies Inc.
    Model:   AMD Radeon R7 200 Series
    Driver:  4.4.12874 Compatibility Profile Context 14.100.0.0

Try updating the driver to the latest version available.
If that fails you might need a different GPU. Sorry!

	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.checkOpenGL(LwjglGraphics.java:284) ~[classes/:na]
	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.initOpenGL(LwjglGraphics.java:187) ~[classes/:na]
	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.postInitialise(LwjglGraphics.java:101) ~[classes/:na]
	at org.terasology.engine.TerasologyEngine.postInitSubsystems(TerasologyEngine.java:266) ~[classes/:na]
	at org.terasology.engine.TerasologyEngine.<init>(TerasologyEngine.java:169) ~[classes/:na]
	at org.terasology.engine.Terasology.main(Terasology.java:118) [classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.7.0_55]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) ~[na:1.7.0_55]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.7.0_55]
	at java.lang.reflect.Method.invoke(Method.java:606) ~[na:1.7.0_55]
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134) [idea_rt.jar:na]
13:27:52.159 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
```

**Comment by Cervator** [comment]
Oh, I did add a tiny tweak for the log statement just thinking about language / tech barrier ("What's a GPU?") and in the rare case a user has two GPUs and the option to pick which one to use

https://github.com/MovingBlocks/Terasology/commit/e094862c5727156dbfe13a902664830042e531d9

**Comment by emanuele3d** [comment]
Thanks for the merge Cervator. The rewording is fine, although now I wonder what glGetString(GL_RENDERER) returns in case of two (or more!!!) graphic cards. 

(30 mins later) Hmmm... looks like it's going to return the model of the card in use by the current context, no mention of any other device. I found an *OpenCL*-based example that finds all the available computing devices (CPUs and GPUs) and chooses the first GPU, but I don't think it's worth going down this road at this stage.

**Comment by Cervator** [comment]
Yeah as my logging indicates you just get one. I've got two :-)

Admittedly they're identical, but .. still.
